### ISSUE TITLE
A way to know if a resource are already in the cache

### ISSUE THREAD

**Issue by eduardotd** [isoos/pwa#7]
First of all, thank you very much for this amazing module.

We are working on a project that will cache  documents for use in off line mode.
  .pdfs, .docs, and so on.

Is there a way to know when the resources listed in the additionalUrls list are loaded?

We want to inform the user that he/she can disconect.

Thanks for any direction..

**Comment by isoos** [comment]
@eduardotd: Could you please expand a bit more about your use case? e.g. how does the user select which pdf to cache offline?

**Comment by eduardotd** [comment]
We run a query in our database to select the documents that the user will use in offline mode:
  - Discipline (Mechanical, Eletricity, ...) 
  - State of the document (Approved, Certified. Aproved internally)
This query returns a list of urls that we feed to  the addicionalUrls List.

  List<String> additionalUrls = new List.from([
  // the result of the query here
  ]);

  worker.offlineUrls = new List.from(offline.offlineUrls)
    ..addAll(additionalUrls);

The problem is that this load can take some time and the user must wait until we are certain that the cache is complete...

**Comment by isoos** [comment]
The `additionalUrls` field was meant for caching the application's static assets that are not part of the `pub build`, e.g. external CSS files or similar assets. It is easy to trigger cache invalidation and new cache building for these kind of assets (e.g. any new software version will nuke this cache), and it will be a very bad user experience if you rely on caching your documents that way.

Instead, I'd suggest a short- and a long-term solution. In both case, I suggest you to not use user-specific URLs in your `pwa.dart` to populate your cache, rather trigger it from your web app.

Short-term (which is sub-optimal, but should work now):

1. Create a DynamicCache and router like described here, and size it to your needs: https://github.com/isoos/pwa/tree/master/examples/custom_routes

2. In your client app issue a request for these URLs, e.g. with the `http` package. When these request receive the data, the file will be in the cache.

Long-term:

I'm currently working on a messaging layer that makes it easier to do request-reply invoke patterns between your web app and the PWA service worker. If that is done, you shall be able to initiate cache population in a form like `await client.precache('cache-id', ['http://your-url/']);` making it efficient, and also allowing more dynamism.

Does this give you enough to get it working (with the short-term solution for now)? I'll ping this issue once the messaging is done.

**Comment by eduardotd** [comment]
Thank you very much for the suggestion.

I will close this issue and try to use DynamicCache.

Cheers
